### PR TITLE
 [MAINTENANCE] Move metrics computations to top-level ParameterBuilder

### PR DESCRIPTION
**Pull request by alexsherstinsky**
### Scope -- https://superconductive.atlassian.net/browse/GE-309:
* Nice refactor, whereby get_metric() and get_metrics() is available to any ParameterBuilder
* These methods also handle parameter name expansion (hence, using the $ syntax is allowed)

Please annotate your PR title to describe what the PR does, then give a brief bulleted description of your PR below. PR titles should begin with [BUGFIX], [FEATURE],  [DOCS], or [MAINTENANCE]. If a new feature introduces breaking changes for the Great Expectations API or configuration files, please also add [BREAKING]. You can read about the tags in our [contributor checklist](https://docs.greatexpectations.io/en/latest/contributing/contribution_checklist.html).

Changes proposed in this pull request:
-
-
-

After submitting your PR, CI checks will run and @tiny-tim-bot will check for your CLA signature.

For a PR with nontrivial changes, we review with both design-centric and code-centric lenses.

In a design review, we aim to ensure that the PR is consistent with our relationship to the open source community, with our software architecture and abstractions, and with our users' needs and expectations. That review often starts well before a PR, for example in github issues or slack, so please link to relevant conversations in notes below to help reviewers understand and approve your PR more quickly (e.g. `closes #123`).

Previous Design Review notes:
-
-
-

### Definition of Done
Please delete options that are not relevant.

- [ ] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/en/latest/contributing/style_guide.html?highlight=style%20guide)
- [ ] I have performed a [self-review](https://docs.greatexpectations.io/en/latest/contributing/contribution_checklist.html?highlight=checklist) of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added [unit tests](https://docs.greatexpectations.io/en/latest/contributing/testing.html#contributing-testing-writing-unit-tests) where applicable and made sure that new and existing tests are passing.
- [ ] I have run any local integration tests and made sure that nothing is broken.

Thank you for submitting!

### CODE REVIEW

File: great_expectations/rule_based_profiler/parameter_builder/metric_parameter_builder.py [hunk 1]
```diff
@@ -74,41 +71,28 @@ def _build_parameters(
 
         batch_id: str = self.get_batch_id(variables=variables)
 
-        metric_domain_kwargs: dict = build_metric_domain_kwargs(
+        metric_computation_result: Dict[
+            str,
+            Union[
+                Union[int, np.int32, np.int64, float, np.float32, np.float64],
```
Comment (cdkini) [inline great_expectations/rule_based_profiler/parameter_builder/metric_parameter_builder.py]: I don't think this is a big deal compared to the `Hashable` matter we previously dealt with but is there any type that encapsulates numeric values? Definitely fine as is though.

File: great_expectations/rule_based_profiler/parameter_builder/numeric_metric_range_multi_batch_parameter_builder.py [hunk 2]
```diff
@@ -159,50 +157,32 @@ def _build_parameters(
                 message=f"Utilizing a {self.__class__.__name__} requires a non-empty list of batch identifiers."
             )
 
-        metric_domain_kwargs: dict = build_metric_domain_kwargs(
-            batch_id=None,
+        metric_computation_result: Dict[
```
Comment (cdkini) [inline great_expectations/rule_based_profiler/parameter_builder/numeric_metric_range_multi_batch_parameter_builder.py:168]: Continuing off of the last comment, is there any way we can simplify this type hint? Perhaps a case for defining our own type? What you currently have is representative of the actual object type so there isn't any issue here but it does seem quite complex.

File: great_expectations/rule_based_profiler/parameter_builder/parameter_builder.py [hunk 3]
```diff
@@ -148,6 +157,164 @@ def get_batch_id(
...
+                metric_value = 0.0
+
+        return {
+            "value": metric_value,
+            "details": {
+                "metric_configuration": metric_configuration_arguments,
+            },
+        }
+
+    def get_metrics(
```
Comment (cdkini) [inline great_expectations/rule_based_profiler/parameter_builder/parameter_builder.py:239]: I know we briefly discussed yesterday but if this method is only used for multibatch parameter builders, does it make sense to allow all parameter builders to have access to it?

My original thought was an additional ABC `MultiBatchParameterBuilder` that extends `ParameterBuilder` but I could see that adding an additional layer of complexity we might not want. Maybe a docstring to clarify the usage would be helpful?

**Review by cdkini — COMMENTED** [review]
Moving those methods to the ABC definitely cleans things up nicely! Left a few general comments.

Should the deletion of `simple_dateformat_string_parameter_builder.py` fit this PR? I agree with its removal if it doesn't currently fit with the current state of the Rules Based Profiler but I do think it might be better suited for a PR around cleanup of the codebase.

File: great_expectations/rule_based_profiler/parameter_builder/parameter_builder.py [hunk 4]
```diff
@@ -148,6 +157,164 @@ def get_batch_id(
...
+        )
+        if enforce_numeric:
+            if not is_numeric(value=metric_value):
+                raise ge_exceptions.ProfilerExecutionError(
+                    message=f"""Applicability of {self.__class__.__name__} is restricted to numeric-valued metrics \
+(value of type "{str(type(metric_value))}" was computed).
+"""
+                )
+            if np.isnan(metric_value):
+                if not fill_nan_with_zero:
```
Comment (talagluck) [inline great_expectations/rule_based_profiler/parameter_builder/parameter_builder.py]: I know you have used `fill_nan_with_zero` elsewhere. I think the name is a little confusing. Could we change it to `replace_nan_with_zero`?

File: great_expectations/rule_based_profiler/parameter_builder/parameter_builder.py (hunk 3 above)
Comment (talagluck) [inline great_expectations/rule_based_profiler/parameter_builder/parameter_builder.py:239]: Yes - I'm also a little confused by `get_metric` vs `get_metrics`. It appears that there is a lot of overlap, and I wonder if we could simplify one or the other.

File: great_expectations/rule_based_profiler/parameter_builder/parameter_builder.py [hunk 5]
```diff
@@ -148,6 +157,164 @@ def get_batch_id(
...
+                    metric_value = 0.0
+
+            metric_values.append(metric_value)
+
+        return {
+            "value": metric_values,
+            "details": {
+                "metric_configuration": {
+                    "metric_name": metric_name,
+                    "domain_kwargs": domain_kwargs,
```
Comment (talagluck) [inline great_expectations/rule_based_profiler/parameter_builder/parameter_builder.py:326]: Why the change from `metric_domain_kwargs` to `domain_kwargs`?

File: great_expectations/rule_based_profiler/parameter_builder/parameter_builder.py [hunk 6]
```diff
@@ -148,6 +157,164 @@ def get_batch_id(
...
         return batch_ids[0]
 
+    def get_metric(
+        self,
+        batch_id: str,
+        validator: Validator,
+        metric_name: str,
+        metric_domain_kwargs: Optional[dict] = None,
+        metric_value_kwargs: Optional[dict] = None,
+        enforce_numeric: Optional[bool] = False,
```
Comment (talagluck) [inline great_expectations/rule_based_profiler/parameter_builder/parameter_builder.py]: Could you explain `enforce_numeric` as well as the assumption that metrics will likely be numeric? I understand this in the NumericRangeParameterBuilder, but not in this parent class, where we might expect metrics which are not numeric.

File: great_expectations/rule_based_profiler/parameter_builder/metric_parameter_builder.py [hunk 7]
```diff
@@ -74,41 +71,28 @@ def _build_parameters(
...
+                Dict[str, Any],
+            ],
+        ] = self.get_metric(
             batch_id=batch_id,
+            validator=validator,
+            metric_name=self._metric_name,
             metric_domain_kwargs=self._metric_domain_kwargs,
+            metric_value_kwargs=self._metric_value_kwargs,
+            fill_nan_with_zero=True,
```
Comment (talagluck) [inline great_expectations/rule_based_profiler/parameter_builder/metric_parameter_builder.py]: Do we want to hard-code `fill_nan_with_zero` here? If someone requests the `column.value_set` metric, and the column is filled with NaNs, what will be returned here?

**Review by talagluck — COMMENTED** [review]
This looks great, though I have a bunch of questions. A walkthrough might be helpful.

File: great_expectations/rule_based_profiler/parameter_builder/parameter_builder.py (hunk 4 above)
Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/parameter_builder/parameter_builder.py]: This is handled consistently now.

File: great_expectations/rule_based_profiler/parameter_builder/parameter_builder.py (hunk 3 above)
Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/parameter_builder/parameter_builder.py:239]: We now add number of batches to metadata.

File: great_expectations/rule_based_profiler/parameter_builder/parameter_builder.py (hunk 6 above)
Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/parameter_builder/parameter_builder.py]: Handled as discussed.

File: great_expectations/rule_based_profiler/parameter_builder/metric_parameter_builder.py (hunk 7 above)
Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/parameter_builder/metric_parameter_builder.py]: Now a parameter in `YML`.

File: great_expectations/rule_based_profiler/parameter_builder/numeric_metric_range_multi_batch_parameter_builder.py (hunk 2 above)
Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/parameter_builder/numeric_metric_range_multi_batch_parameter_builder.py:168]: This has been resolved!  Yay!

File: great_expectations/rule_based_profiler/parameter_builder/metric_parameter_builder.py (hunk 1 above)
Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/parameter_builder/metric_parameter_builder.py]: We have a better way now!  Thanks for causing us to think this over!

File: great_expectations/rule_based_profiler/parameter_builder/metric_parameter_builder.py [hunk 8]
```diff
@@ -29,6 +25,8 @@ def __init__(
         metric_name: str,
         metric_domain_kwargs: Optional[Union[str, dict]] = None,
         metric_value_kwargs: Optional[Union[str, dict]] = None,
+        enforce_numeric_metric: Optional[Union[str, bool]] = False,
+        fill_nan_with_zero: Optional[Union[str, bool]] = False,
```
Comment (cdkini) [inline great_expectations/rule_based_profiler/parameter_builder/metric_parameter_builder.py]: In what conditions would `enforce_numeric_metric` or `fill_nan_with_zero` be a `str`? Is it in the case we use `$variable` syntax in the config YAML? 

These very much read like boolean arguments so I just want to wrap my head around them.

File: great_expectations/rule_based_profiler/parameter_builder/numeric_metric_range_multi_batch_parameter_builder.py (hunk 2 above)
Comment (cdkini) [reply great_expectations/rule_based_profiler/parameter_builder/numeric_metric_range_multi_batch_parameter_builder.py:168]: Could `Union[Union[np.ndarray, List[Number]], Dict[str, Any]]` be simplified to `Union[np.ndarray, List[Number], Dict[str, Any]]`? Or are `np.ndarray` and `List[Number]` grouped together because they're similar?

File: great_expectations/rule_based_profiler/parameter_builder/parameter_builder.py (hunk 3 above)
Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/parameter_builder/parameter_builder.py:239]: The line between single-batch and multi-batch will soon get very blurred as use cases will progressively weigh in favor of multi-batch, while single-batch will be a special case of multi-batch.

File: great_expectations/rule_based_profiler/parameter_builder/parameter_builder.py [hunk 9]
```diff
@@ -148,6 +158,179 @@ def get_batch_id(
 
         return batch_ids[0]
 
+    def get_numeric_metric(
```
Comment (cdkini) [inline great_expectations/rule_based_profiler/parameter_builder/parameter_builder.py]: If we have `get_numeric_metric` in the ABC, is the plan to eventually add other getters for other metric types in this same parent definition?

File: great_expectations/rule_based_profiler/parameter_builder/parameter_builder.py [hunk 10]
```diff
@@ -148,6 +158,179 @@ def get_batch_id(
...
+        metric_configuration_arguments: Dict[str, Any] = {
+            "metric_name": metric_name,
+            "metric_domain_kwargs": metric_domain_kwargs,
+            "metric_value_kwargs": metric_value_kwargs,
+            "metric_dependencies": None,
+        }
+        metric_value: Number = validator.get_metric(
+            metric=MetricConfiguration(**metric_configuration_arguments)
+        )
+        if enforce_numeric_metric:
```
Comment (cdkini) [inline great_expectations/rule_based_profiler/parameter_builder/parameter_builder.py:217]: Could this and L224 combine their conditionals into a single line? Very much a stylistic preference without impact on code so I wouldn't dwell on it if you want to merge.

`i.e. if enforce_numeric_metric and is not is_numeric:`

**Review by cdkini — COMMENTED** [review]
Added a few comments/questions after the last review

File: great_expectations/rule_based_profiler/parameter_builder/metric_parameter_builder.py (hunk 8 above)
Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/parameter_builder/metric_parameter_builder.py]: @cdkini They are strings because of the string-based `$parameter` / `$variable` usage in `YML` configuration.

File: great_expectations/rule_based_profiler/parameter_builder/parameter_builder.py (hunk 9 above)
Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/parameter_builder/parameter_builder.py]: @cdkini Potentially, only as the needs arise.

File: great_expectations/rule_based_profiler/parameter_builder/numeric_metric_range_multi_batch_parameter_builder.py [hunk 11]
```diff
@@ -46,6 +47,8 @@ def __init__(
         metric_name: str,
         metric_domain_kwargs: Optional[Union[str, dict]] = None,
         metric_value_kwargs: Optional[Union[str, dict]] = None,
+        enforce_numeric_metric: Optional[Union[str, bool]] = False,
```
Comment (talagluck) [inline great_expectations/rule_based_profiler/parameter_builder/numeric_metric_range_multi_batch_parameter_builder.py]: I could see setting this to True for this parameter builder.

File: great_expectations/rule_based_profiler/parameter_builder/parameter_builder.py (hunk 3 above)
Comment (cdkini) [reply great_expectations/rule_based_profiler/parameter_builder/parameter_builder.py:239]: Sounds good, thanks for the clarification! I'll try and update my mental model for the future.

File: great_expectations/rule_based_profiler/parameter_builder/metric_parameter_builder.py (hunk 8 above)
Comment (cdkini) [reply great_expectations/rule_based_profiler/parameter_builder/metric_parameter_builder.py]: Makes a lot of sense, thanks for the clarification.

File: great_expectations/rule_based_profiler/parameter_builder/parameter_builder.py (hunk 10 above)
Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/parameter_builder/parameter_builder.py:217]: @cdkini I prefer to leave it as is, because `enforce_numeric_metric` is common (hence, should be factored out and not evaluated twice, regardless of how inexpensive that would be).

File: great_expectations/rule_based_profiler/parameter_builder/parameter_builder.py (hunk 5 above)
Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/parameter_builder/parameter_builder.py:326]: Added number of batches to `metadata`.

**Review by talagluck — COMMENTED** [review]
LGTM - almost there! As discussed:

- Please change the default value for `enforce_numeric_metric` 
- Please change `get_numeric_metric`, etc. to `get_metric` and ensure that the return types are `Any` instead of `Numeric`.
- Please change `fill_nan_with_zero` to `replace_nan_with_zero`

File: great_expectations/rule_based_profiler/parameter_builder/numeric_metric_range_multi_batch_parameter_builder.py (hunk 11 above)
Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/parameter_builder/numeric_metric_range_multi_batch_parameter_builder.py]: Change made.

File: great_expectations/rule_based_profiler/parameter_builder/parameter_builder.py (hunk 4 above)
Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/parameter_builder/parameter_builder.py]: Chanted to `replace_nan_with_zero` per our discussion.

**Review by talagluck — APPROVED** [review]
LGTM! Thanks for the changes
